sdk/python/lib/pulumi/output.py

```diff
@@ -139,6 +144,68 @@ def is_known(self) -> Awaitable[bool]:
 
     # End private implementation details.
 
+    def __getstate__(self):
+        """
+        Serialize this Output into a dictionary for pickling, only when serialization is enabled.
+        """
+
+        if not _serialization_enabled():
+            raise Exception("__getstate__ can only be called during serialization")
+
+        value, is_secret = _sync_await(asyncio.gather(self.future(), self.is_secret()))
+
+        if is_secret:
+            if _secrets_allowed():
+                _set_contained_secrets(True)
+            else:
+                raise Exception("Secret outputs cannot be captured")
+
+        return {"value": value}
+
+    def __setstate__(self, state):
+        """
+        Deserialize this Output from a dictionary, only when serialization is enabled.
+        """
+
+        if not _serialization_enabled():
+            raise Exception("__setstate__ can only be called during deserialization")
+
+        value = state["value"]
+
+        # Replace '.get' with a function that returns the value without raising an error.
+        self.get = lambda: value
+
+        def error(name: str):
+            def f(*args: Any, **kwargs: Any):
+                raise Exception(
+                    f"'{name}' is not allowed from inside a cloud-callback. "
+                    + "Use 'get' to retrieve the value of this Output directly."
+                )
+
+            return f
+
+        # Replace '.apply' and other methods on Output with implementations that raise an error.
+        self.apply = error("apply")
+        self.resources = error("resources")
+        self.future = error("future")
+        self.is_known = error("is_known")
+        self.is_secret = error("is_secret")
+
+    def get(self) -> T_co:
+        """
+        Retrieves the underlying value of this Output.
+
+        This function is only callable in code that runs in the cloud post-deployment. At this
```

Comment:
I copied this text from Node.js SDK. Technically, we don't have full closure serialization in Python so the "in the cloud" part isn't really accurate. I could strip that part here. This is only really available to be called from a deserialized dynamic `ResourceProvider` subclass.